### PR TITLE
ARROW-16219: [CI] Fix git config to prevent SCM tools failure

### PR DESCRIPTION
**Pull request by raulcd**
This PR fixes the CI failures due to the latest git release fixing CVE-2022-24765.
I have been able to see the build passing the scm step with the change here:
https://app.travis-ci.com/github/raulcd/arrow/builds/249688925
The above build fails due to some tests failing but not related with installation anymore.

And this was the failure before the change:
https://app.travis-ci.com/github/raulcd/arrow/builds/249683847

I also have been able to reproduce the issue on the `verify-conda-rc` locally.

The failure:
  ```
  $ docker-compose run  conda-verify-rc

....
  Preparing transaction: done
Verifying transaction: done
Executing transaction: done
/arrow/python /arrow /
Traceback (most recent call last):
  File "/arrow/python/setup.py", line 607, in <module>
    setup(
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/__init__.py", line 87, in setup
    return distutils.core.setup(**attrs)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 109, in setup
    _setup_distribution = dist = klass(attrs)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/dist.py", line 462, in __init__
    _Distribution.__init__(
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 293, in __init__
    self.finalize_options()
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/dist.py", line 886, in finalize_options
    ep(self)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools/dist.py", line 907, in _finalize_setup_keywords
    ep.load()(self, ep.name, value)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools_scm/integration.py", line 75, in version_keyword
    _assign_version(dist, config)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools_scm/integration.py", line 51, in _assign_version
    _version_missing(config)
  File "/tmp/arrow-HEAD.YUVPq/mambaforge/envs/conda-source/lib/python3.10/site-packages/setuptools_scm/__init__.py", line 106, in _version_missing
    raise LookupError(
LookupError: setuptools-scm was unable to detect version for /arrow.

Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.

For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
Failed to verify release candidate. See /tmp/arrow-HEAD.YUVPq for details.
  ```

Waiting to validate the verify-rc fix at the moment, will update once the local build finishes

### CODE REVIEW

**Comment by raulcd** [comment]
We can see on the new build how the `ARROW_GIT_ID` variable is correctly populated now too:
```
--   ARROW_GIT_ID=f94a14ddd3a5271a3779b831da482dfe212d57a7 [default=""]
--       The Arrow git commit id (if any)
```

**Comment by raulcd** [comment]
Reproduced the issue and fixed on verify-release-candidate too:
```
$ docker-compose run  conda-verify-rc
...
Executing transaction: done
/arrow/python /arrow /
running build_ext
-- Running cmake for pyarrow
cmake -DPYTHON_EXECUTABLE=/tmp/arrow-HEAD.fNS4t/mambaforge/envs/conda-source/bin/python -DPython3_EXECUTABLE=/tmp/arrow-HEAD.fNS4t/mambaforge/envs/conda-source/bin/python "" -DPYARROW_BUILD_CUDA=off -DPYARROW_BUILD_FLIGHT=on -DPYARROW_BUILD_GANDIVA=on -DPYARROW_BUILD_DATASET=on -DPYARROW_BUILD_ORC=on -DPYARROW_BUILD_PARQUET=on -DPYARROW_BUILD_PARQUET_ENCRYPTION=on -DPYARROW_BUILD_PLASMA=on -DPYARROW_BUILD_S3=off -DPYARROW_BUILD_HDFS=on -DPYARROW_USE_TENSORFLOW=off -DPYARROW_BUNDLE_ARROW_CPP=off -DPYARROW_BUNDLE_BOOST=off -DPYARROW_GENERATE_COVERAGE=off -DPYARROW_BOOST_USE_SHARED=on -DPYARROW_PARQUET_USE_SHARED=on -DCMAKE_BUILD_TYPE=release /arrow/python
CMake Warning:
  Ignoring empty string ("") provided on the command line.


-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /tmp/arrow-HEAD.fNS4t/mambaforge/envs/conda-source/bin/x86_64-conda-linux-gnu-cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /tmp/arrow-HEAD.fNS4t/mambaforge/envs/conda-source/bin/x86_64-conda-linux-gnu-c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- System processor: x86_64
-- Performing Test CXX_SUPPORTS_SSE4_2
-- Performing Test CXX_SUPPORTS_SSE4_2 - Success
-- Performing Test CXX_SUPPORTS_AVX2
-- Performing Test CXX_SUPPORTS_AVX2 - Success
-- Performing Test CXX_SUPPORTS_AVX512
-- Performing Test CXX_SUPPORTS_AVX512 - Success
-- Arrow build warning level: PRODUCTION
Using ld linker

```

**Comment by raulcd** [comment]
@jorisvandenbossche @pitrou I've been able to validate on travis for `s390x Python` and locally for `conda-verify-rc` the fix as shown on the comments. We probably want to wait until the build passes but should be ready for review.

File: dev/release/verify-release-candidate.sh [hunk 1]
```diff
@@ -895,6 +895,8 @@ ensure_source_directory() {
     fi
   fi
 
+  # Ensure ARROW_SOURCE_DIR is a safe.directory for git
+  git config --global --add safe.directory $ARROW_SOURCE_DIR
```
Comment (pitrou) [inline dev/release/verify-release-candidate.sh:899]: Hmm, this script can be run by developers on their local machine, can we run the `git config` only on CI configs? (for example in the parent script or yaml file)

Comment (raulcd) [reply dev/release/verify-release-candidate.sh:899]: We can add it to the docker-compose.yml:
```
  conda-verify-rc:
    image: ubuntu:${UBUNTU}
    volumes:
      - .:/arrow:delegated
      - ${DOCKER_VOLUME_PREFIX}conda-ccache:/ccache:delegated
    shm_size: '1gb'
    environment:
      <<: *ccache
      CMAKE_GENERATOR: Ninja
      DEBIAN_FRONTEND: "noninteractive"
      DOTNET_SYSTEM_GLOBALIZATION_INVARIANT: 1
      TEST_APT: 0  # would require docker-in-docker
      TEST_YUM: 0
      USE_CONDA: 1
    command: >
      /bin/bash -c "
        apt update -y && apt install -y curl git gnupg tzdata wget &&
        git config --global --add safe.directory /arrow &&
        /arrow/dev/release/verify-release-candidate.sh $${VERIFY_VERSION} $${VERIFY_RC}"
```
If we want to add it on the CI files only (not even `docker-compose.yml` we can modify the archery call `arrow/dev/tasks/verify-rc/github.linux.amd64.docker.yml` and add an environment variable to manage it:
```
      - name: Execute Docker Build
        shell: bash
        run: |
          archery docker run \
            -e VERIFY_VERSION="{{ release|default("") }}" \
            -e VERIFY_RC="{{ rc|default("") }}" \
            -e TEST_DEFAULT=0 \
            -e TEST_{{ target|upper }}=1 \
            -e SET_GIT_SAFE_DIR=1 \
            {{ distro }}-verify-rc
```
Would you be ok with modifying the docker-compose as shown above?

**Comment by raulcd** [comment]
@github-actions crossbow submit verify-rc-source-python-linux-conda-latest-amd64

**Comment by kou** [comment]
Can we add `-c safe.directory=...` to `git` command lines in `cpp/cmake_modules/DefineOptions.cmake` and `python/setup.py` instead of adding `git config --global`?

**Comment by pitrou** [comment]
One problem is that AFAICT a git command is implicitly issued by setuptools_scm.

**Comment by pitrou** [comment]
@kou Apparently one cannot use `-c safe.directory`: https://git-scm.com/docs/git-config/2.36.0#Documentation/git-config.txt-safedirectory

> This config setting is only respected when specified in a system or global config, not when it is specified in a repository config or via the command line option -c safe.directory=<path>.

**Comment by kou** [comment]
> @kou Apparently one cannot use `-c safe.directory`: https://git-scm.com/docs/git-config/2.36.0#Documentation/git-config.txt-safedirectory
> 
> > This config setting is only respected when specified in a system or global config, not when it is specified in a repository config or via the command line option -c safe.directory=.

Oh, sorry. I missed this.

**Comment by raulcd** [comment]
About the build failures, I am going to fix the Windows ones which are introduced by this PR https://github.com/apache/arrow/runs/6114986154?check_suite_focus=true:
```
Run export CMAKE_BUILD_PARALLEL_LEVEL=$NUMBER_OF_PROCESSORS
+ source_dir=/d/a/arrow/arrow/cpp
+ build_dir=/d/a/arrow/arrow/build/cpp
+ : ON
+ : OFF
+ git config --global --add safe.directory /d/a/arrow/arrow
ci/scripts/cpp_build.sh: line 28: git: command not found
```
The MacOS ones are related to homebrew installing git https://github.com/apache/arrow/runs/6114984587?check_suite_focus=true:
```
 Error: Could not symlink share/doc/git-doc/MyFirstContribution.html
Target /usr/local/share/doc/git-doc/MyFirstContribution.html
is a symlink belonging to git@2.35.1. You can unlink it:
  brew unlink git@2.35.1

To force the link and overwrite all conflicting files:
  brew link --overwrite git@2.35.1

To list all files that would be deleted:
  brew link --overwrite --dry-run git@2.35.1
Installing git has failed!
```

**Comment by raulcd** [comment]
@kou this PR fixes all the git failures I've found so far except the brew installation that I've captured on a different ticket (https://issues.apache.org/jira/browse/ARROW-16278) and have started working on a different PR https://github.com/apache/arrow/pull/12956 . Let me know if you think we should do other changes or if I should incorporate the brew fixes here (I am still working on understanding how to fix them).

**Comment by raulcd** [comment]
@kszucs was trying a different approach which seems better, closing this PR in favor of the other one (https://github.com/apache/arrow/pull/12958)

**Comment by raulcd** [comment]
Sorry, this was not the PR to close

**Review by kszucs — APPROVED** [review]
Thanks @raulcd! Merging on green.

**Pull request closed without merging**